atlasdb-commons/src/test/java/com/palantir/common/concurrent/ConcurrentStreamsTest.java

```diff
@@ -54,30 +57,69 @@ public void testDoesEvaluateResultsWhenLimitingConcurrency() {
     }
 
     @Test
+    public void testDoesMaintainCorrectOrdering() {
+        Stream<Integer> values = ConcurrentStreams.map(
+                ImmutableList.of(5, 3, 6, 2, 1, 9),
+                value -> {
+                    if (value <= 3) {
+                        pause(100);
+                    }
+                    return value + 1;
+                },
+                executor,
+                2);
+        Assert.assertEquals(values.collect(Collectors.toList()), ImmutableList.of(6, 4, 7, 3, 2, 10));
+    }
+
+    @Test
+    public void testCanHandleValueDuplicates() {
+        Stream<Integer> values = ConcurrentStreams.map(
+                ImmutableList.of(1, 2, 1, 2),
+                value -> value + 1,
+                executor,
+                2);
+        Assert.assertEquals(values.collect(Collectors.toList()), ImmutableList.of(2, 3, 2, 3));
+    }
+
+    @Test
     public void testShouldOnlyRunWithProvidedConcurrency() throws Exception {
+        CountDownLatch stage1 = new CountDownLatch(3);
```

Comment:
Can we make the countdown names more descriptive or extract methods or add comments. I dont think I understand the intent behind each latch, but am pretty sure its has been well thought out :p. Similarly for other tests.